lntest/itest/lnd_max_htlcs_test.go

```diff
@@ -0,0 +1,215 @@
+	maxHtlcs := 5
+
+	ctxt, _ := context.WithTimeout(ctxb, channelOpenTimeout)
+	chanPoint := openChannelAndAssert(
+		ctxt, t, net, net.Alice, net.Bob,
+		lntest.OpenChannelParams{
+			Amt:            1000000,
+			PushAmt:        800000,
+			RemoteMaxHtlcs: uint16(maxHtlcs),
+		},
+	)
+
+	// Wait for Alice and Bob to receive the channel edge from the
+	// funding manager.
+	ctxt, _ = context.WithTimeout(ctxb, defaultTimeout)
+	err := net.Alice.WaitForNetworkChannelOpen(ctxt, chanPoint)
+	require.NoError(t.t, err, "alice does not have open channel")
+
+	ctxt, _ = context.WithTimeout(ctxb, defaultTimeout)
+	err = net.Bob.WaitForNetworkChannelOpen(ctxt, chanPoint)
+	require.NoError(t.t, err, "bob does not have open channel")
+
+	// Alice and bob should have one channel open with each other now.
+	assertNodeNumChannels(t, net.Alice, 1)
+	assertNodeNumChannels(t, net.Bob, 1)
+
+	// Send our maximum number of htlcs from Bob -> Alice so that we get
+	// to a point where Alice won't accept any more htlcs on the channel.
+	subscriptions := make([]*holdSubscription, maxHtlcs)
+	subCtx, cancel := context.WithTimeout(ctxb, defaultTimeout)
+	defer cancel()
+
+	for i := 0; i < maxHtlcs; i++ {
+		subscriptions[i] = acceptHoldInvoice(
+			subCtx, t.t, i, net.Bob, net.Alice,
+		)
+	}
+
+	err = assertNumActiveHtlcs([]*lntest.HarnessNode{
+		net.Alice, net.Bob,
+	}, maxHtlcs)
+	require.NoError(t.t, err, "htlcs not active")
+
+	// Now we send a payment from Alice -> Bob to sanity check that our
+	// commitment limit is not applied in the opposite direction.
+	aliceBobSub := acceptHoldInvoice(
+		subCtx, t.t, maxHtlcs, net.Alice, net.Bob,
+	)
+	err = assertNumActiveHtlcs([]*lntest.HarnessNode{
+		net.Alice, net.Bob,
+	}, maxHtlcs+1)
+	require.NoError(t.t, err, "htlcs not active")
+
+	// Now, we're going to try to send another payment from Bob -> Alice.
+	// We've hit our max remote htlcs, so we expect this payment to spin
+	// out dramatically with pathfinding.
+	ctxt, _ = context.WithTimeout(ctxb, defaultTimeout)
+	payment, err := net.Bob.RouterClient.SendPaymentV2(
+		ctxt, &routerrpc.SendPaymentRequest{
+			Amt:            1000,
+			Dest:           net.Alice.PubKey[:],
+			TimeoutSeconds: 60,
+			FeeLimitSat:    1000000,
+			MaxParts:       10,
+			Amp:            true,
+		},
+	)
+	require.NoError(t.t, err, "send payment failed")
+
+	update, err := payment.Recv()
+	require.NoError(t.t, err, "no payment in flight update")
+	require.Equal(t.t, lnrpc.Payment_IN_FLIGHT, update.Status,
+		"payment not inflight")
+
+	update, err = payment.Recv()
+	require.NoError(t.t, err, "no payment failed update")
+	require.Equal(t.t, lnrpc.Payment_FAILED, update.Status)
+	require.Len(t.t, update.Htlcs, 0, "expected no htlcs dispatched")
+
+	// Now that we're done, we cancel all our pending htlcs so that we
+	// can cleanup the channel with a coop close.
+	ctxt, _ = context.WithTimeout(ctxb, defaultTimeout)
+	for _, sub := range subscriptions {
+		sub.cancel(ctxt, t.t)
```

Comment:
nit: think this could use distinct `ctxt` so that all these `cancel` calls aren't under the same `defaultTimeout`?